### PR TITLE
ENH: scipy.signal add firls and tests

### PR DESCRIPTION
**Pull request by GregoryEAllen**
As discussed in the mailing list, plus tests

### CODE REVIEW

File: scipy/signal/fir_filter_design.py [hunk 1]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
     bands = np.asarray(bands).copy()
     return sigtools._remez(numtaps, bands, desired, weight, tnum, Hz,
                            maxiter, grid_density)
+
+
+def firls(numtaps, bands, desired, weight=None, Hz=1):
```
Comment (WarrenWeckesser) [inline scipy/signal/fir_filter_design.py]: Using `Hz` as the argument name is consistent with `remez`, but it's not a great name for this argument.  This is a case where I think we should use a better name instead of using a bad name for the sake of consistency.  `coherence`, `csd`, `periodogram`, and `welch` use `fs` for the sampling frequency, while `firwin` and `firwin2` use `nyq` for the Nyquist frequency.  My preference is `fs`, but either is better than `Hz`.

File: scipy/signal/fir_filter_design.py [hunk 2]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+        All elements must be non-negative and less than half the sampling
+        frequency as given by `Hz`.
+    desired : array_like
+        A sequence half the size of bands containing the desired gain
+        in each of the specified bands.
+    weight : array_like, optional
+        A relative weighting to give to each band region. The length of
+        `weight` has to be half the length of `bands`.
+    Hz : scalar, optional
+        The sampling frequency in Hz. Default is 1.
```
Comment (WarrenWeckesser) [inline scipy/signal/fir_filter_design.py]: This should not say `in Hz`.  There is no requirement that the units of the sampling rate be Hz.

**Comment by WarrenWeckesser** [comment]
In the Matlab (and Octave) version of this function (e.g. http://www.mathworks.com/help/signal/ref/firls.html), the `desired` argument is the same length as the `bands` argument.  It gives the values at the ends of the bands, which allows the desired filter to be described as piecewise linear segments (rather than piecewise constant).  It would be nice to provide the same API here.

File: scipy/signal/fir_filter_design.py [hunk 3]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+
+    Returns
+    -------
+    out : ndarray
+        A rank-1 array containing the coefficients of the optimal
+        (in a least squares sense) filter.
+
+    Example
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: Should be ``Examples`` for it to render correctly as html/pdf.

File: scipy/signal/fir_filter_design.py [hunk 4]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+    >>> ampl = np.abs(response)
+    >>> import matplotlib.pyplot as plt
+    >>> fig = plt.figure()
+    >>> ax1 = fig.add_subplot(111)
+    >>> ax1.semilogy(freq/(2*np.pi), ampl, 'b-')  # freq in Hz
+    >>> plt.show()
+    """
+    if numtaps % 2 == 0:
+        raise ValueError("numtaps must be odd.")
+    L = (numtaps-1)//2
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: pep8 nitpick: spaces around ``//``

File: scipy/signal/fir_filter_design.py [hunk 5]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+    -------
+    We want to construct a filter with a passband at 0.2-0.3 Hz, and
+    stop bands at 0-0.1 Hz and 0.4-0.5 Hz. Note that this means that the
+    behavior in the frequency ranges between those bands is unspecified and
+    may overshoot.
+    >>> from scipy import signal
+    >>> bpass = signal.firls(71, [0, 0.1, 0.2, 0.3, 0.4, 0.5], [0, 1, 0])
+    >>> freq, response = signal.freqz(bpass)
+    >>> ampl = np.abs(response)
+    >>> import matplotlib.pyplot as plt
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: can you move this import to right below the ``signal`` import?

File: scipy/signal/fir_filter_design.py [hunk 6]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+    """
+    if numtaps % 2 == 0:
+        raise ValueError("numtaps must be odd.")
+    L = (numtaps-1)//2
+
+    # normalize bands and make it 2 columns
+    bands = np.asarray(bands).flatten()/Hz
+    if len(bands) % 2 == 1:
+        raise ValueError("bands must contain frequency pairs.")
+    bands = bands.reshape(-1,2)
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: pep8: space after comma

File: scipy/signal/fir_filter_design.py [hunk 7]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+        raise ValueError("desired must have one entry per band.")
+    if weight is None:
+        weight = np.ones_like(desired)
+
+    # set up the linear matrix equation to be solved, Ax = b
+    k = np.arange(L+1)[np.newaxis]
+    m = k.T
+    A,b = 0,0
+    for i, (f0,f1) in enumerate(bands):
+        Ai = f1 * (sinc(2*(m+k)*f1) + sinc(2*(m-k)*f1)) \
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: can you use an extra ``()`` and remove the backslash?

File: scipy/signal/fir_filter_design.py [hunk 8]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+    # check remaining params
+    if len(bands) != len(desired):
+        raise ValueError("desired must have one entry per band.")
+    if weight is None:
+        weight = np.ones_like(desired)
+
+    # set up the linear matrix equation to be solved, Ax = b
+    k = np.arange(L+1)[np.newaxis]
+    m = k.T
+    A,b = 0,0
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: two separate lines

File: scipy/signal/tests/test_fir_filter_design.py [hunk 9]
```diff
@@ -390,5 +390,46 @@ def test_hilbert(self):
...
+        a = 0.1  # width of the transition band
+
+        # design a halfband symmetric low-pass filter
+        h = firls(11, [0, a, 0.5-a, 0.5], [1, 0])
+
+        # make sure the filter has correct # of taps
+        assert_equal(len(h), N)
+
+        # make sure it is symmetric
+        midx = (N-1)//2
```
Comment (rgommers) [inline scipy/signal/tests/test_fir_filter_design.py]: spaces around ``//``

File: scipy/signal/tests/test_fir_filter_design.py [hunk 10]
```diff
@@ -390,5 +390,46 @@ def test_hilbert(self):
...
+        assert_array_almost_equal(h[:midx], h[:-midx-1:-1])
+
+        # make sure the center tap is 0.5
+        assert_almost_equal(h[midx], 0.5)
+
+        # For halfband symmetric, odd coefficients (except the center)
+        # should be zero (really small)
+        hodd = np.hstack((h[1:midx:2],h[-midx+1::2]))
+        assert_array_almost_equal(hodd,0)
```
Comment (rgommers) [inline scipy/signal/tests/test_fir_filter_design.py]: spaces after comma in the 2 lines above

File: scipy/signal/tests/test_fir_filter_design.py [hunk 11]
```diff
@@ -390,5 +390,46 @@ def test_hilbert(self):
...
+
+        # now check the frequency response
+        w, H = freqz(h, 1)
+        f = w/2/np.pi
```
Comment (rgommers) [inline scipy/signal/tests/test_fir_filter_design.py]: more spaces, and maybe clearer: ``w / (2*np.pi)``

File: scipy/signal/tests/test_fir_filter_design.py [hunk 12]
```diff
@@ -390,5 +390,46 @@ def test_hilbert(self):
...
+        Hmag = abs(H)
```
Comment (rgommers) [inline scipy/signal/tests/test_fir_filter_design.py]: ``np.abs``?

File: scipy/signal/tests/test_fir_filter_design.py [hunk 13]
```diff
@@ -390,5 +390,46 @@ def test_hilbert(self):
...
+
+        # check that the pass band is close to unity
+        idx = (f > 0) * (f < a)
+        assert_array_almost_equal(Hmag[idx],1, decimal=3)
```
Comment (rgommers) [inline scipy/signal/tests/test_fir_filter_design.py]: one more space

File: scipy/signal/tests/test_fir_filter_design.py [hunk 14]
```diff
@@ -390,5 +390,46 @@ def test_hilbert(self):
...
+
+        # check that the stop band is close to zero
+        idx = (f > 0.5-a) * (f < 0.5)
```
Comment (rgommers) [inline scipy/signal/tests/test_fir_filter_design.py]: ``np.logical_and`` for clarity?

File: scipy/signal/tests/test_fir_filter_design.py [hunk 15]
```diff
@@ -390,5 +390,46 @@ def test_hilbert(self):
...
+        assert_array_almost_equal(Hmag[idx],0, decimal=3)
```
Comment (rgommers) [inline scipy/signal/tests/test_fir_filter_design.py]: and one more space after comma

File: scipy/signal/fir_filter_design.py [hunk 16]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+        `weight` has to be half the length of `bands`.
+    Hz : scalar, optional
+        The sampling frequency in Hz. Default is 1.
+
+    Returns
+    -------
+    out : ndarray
+        A rank-1 array containing the coefficients of the optimal
+        (in a least squares sense) filter.
+
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: A ``See Also`` section would be useful here.

File: scipy/signal/fir_filter_design.py [hunk 17]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+        in each of the specified bands.
+    weight : array_like, optional
+        A relative weighting to give to each band region. The length of
+        `weight` has to be half the length of `bands`.
+    Hz : scalar, optional
+        The sampling frequency in Hz. Default is 1.
+
+    Returns
+    -------
+    out : ndarray
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: maybe name this return ``coeffs`` or something similar

**Comment by rgommers** [comment]
Can you also add ``firls`` to the function listing in ``signal/__init__.py``? That ensures that it'll be listed in the reference guide.

File: scipy/signal/fir_filter_design.py [hunk 18]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+        A rank-1 array containing the coefficients of the optimal
+        (in a least squares sense) filter.
+
+    Example
+    -------
+    We want to construct a filter with a passband at 0.2-0.3 Hz, and
+    stop bands at 0-0.1 Hz and 0.4-0.5 Hz. Note that this means that the
+    behavior in the frequency ranges between those bands is unspecified and
+    may overshoot.
```
Comment (rgommers) [inline scipy/signal/fir_filter_design.py]: need a blank line here in order for the example to render correctly

**Comment by rgommers** [comment]
I've been playing with this a bit, trying to replicate some Matlab examples. It's not very intuitive that ``desired`` is the gain in a band (is that averated over the whole band?). Matlab uses a gain per point, i.e. ``bands`` and ``desired`` have the same length. Is this difference on purpose?

**Comment by rgommers** [comment]
It would also be good to add a ``Notes`` section that explains exactly which type of ``firls`` filter this is. I.e. even/odd filter, and whether or not there's a boundary condition at the Nyquist frequency.

File: scipy/signal/fir_filter_design.py [hunk 19]
```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
...
+
+    # set up the linear matrix equation to be solved, Ax = b
+    k = np.arange(L+1)[np.newaxis]
+    m = k.T
+    A,b = 0,0
+    for i, (f0,f1) in enumerate(bands):
+        Ai = f1 * (sinc(2*(m+k)*f1) + sinc(2*(m-k)*f1)) \
+           - f0 * (sinc(2*(m+k)*f0) + sinc(2*(m-k)*f0))
+        bi = desired[i] * (2*f1*sinc(2*m*f1) - 2*f0*sinc(2*m*f0))
+        A += Ai * abs(weight[i]**2)
```
Comment (WarrenWeckesser) [inline scipy/signal/fir_filter_design.py]: The weights should not be squared.  I tried comparing the output of this `firls` to Octave's `firls`, and I get agreement if the weights are not squared:

**`firls` from this pull request, but with the weights *not* squared:**

```
In [4]: taps = firls(9, [0, 0.5, 0.55, 1], [1, 0], [1, 2], Hz=2)

In [5]: taps
Out[5]: 
array([-0.00062693, -0.10335445, -0.00981577,  0.31727169,  0.51140943,
        0.31727169, -0.00981577, -0.10335445, -0.00062693])
```

**`firls` from the Octave signal package:**

```
octave:30> taps = firls(8, [0 0.5 0.55 1], [1 1 0 0], [1, 2]);
octave:31> taps
taps =

  -6.26930101730182e-04
  -1.03354450635036e-01
  -9.81576747564301e-03
   3.17271686090449e-01
   5.11409425599933e-01
   3.17271686090449e-01
  -9.81576747564301e-03
  -1.03354450635036e-01
  -6.26930101730182e-04
```

File: scipy/signal/fir_filter_design.py (hunk 18 above)
Comment (FabricioS) [inline scipy/signal/fir_filter_design.py]: It seems to me that the band specifications should either be relative to a default sampling frequency (Fs=1 in the function signature), or absolute but requiring an explicit definition of the sampling frequency, at least in the example.
So maybe we can suppress the Hz mentions.

File: scipy/signal/fir_filter_design.py (hunk 2 above)
Comment (FabricioS) [inline scipy/signal/fir_filter_design.py]: Usually the numerical sampling frequency is 2 and the related Nyquist frequency is 1, at least in the DSP community.

**Comment by rgommers** [comment]
@GregoryEAllen are you planning to address the review comments? It would be great to get this feature merged before the next release.

**Comment by GregoryEAllen** [comment]
Sorry, I'm swamped right now. I'm not real motivated to make major changes. This was supposed to be low-hanging fruit.

File: scipy/signal/fir_filter_design.py (hunk 2 above)
Comment (endolith) [inline scipy/signal/fir_filter_design.py]: elsewhere we've been saying `ordinary frequency (e.g. Hz)` to distinguish from `angular frequency (e.g. rad/s)`

**Comment by Eric89GXL** [comment]
@GregoryEAllen if I find some time, is it okay with you if I squash your two commits and then work off of it to open another PR? I agree this would be a nice one to have in `scipy`.

**Comment by GregoryEAllen** [comment]
Yes, use it as you will.

**Comment by Eric89GXL** [comment]
@GregoryEAllen as I get going on this, do you have a textbook or link you used to come up with the implementation? It would be good to list it if possible.

**Comment by GregoryEAllen** [comment]
@Eric89GXL I googled for a python firls, and found this one in the scipy mailing list. I cleaned it up in the style of signal.remez and posted it back to the list. It was suggested that this should be added to scipy, so I made a pull request and several minor recommended changes. What tripped me up was disagreement on what the public API should be, i.e. it's done in the style of scipy.remez instead of Matlab/Octave firls. Perhaps scipy should also have firpm and migrate away from remez for consistency across the platforms.
TL/DR: No, it was found in the mailing list.

**Comment by Eric89GXL** [comment]
Closing for #5989
